spec/lang/machine.md

```diff
@@ -321,11 +333,15 @@ impl<M: Memory> Machine<M> {
     }
 
     /// Look up a function given a pointer.
-    fn fn_from_ptr(&self, ptr: ThinPointer<M::Provenance>) -> Result<Function> {
-        let Some((func_name, _)) = self.fn_ptrs.iter().find(|(_, fn_ptr)| *fn_ptr == ptr) else {
-            throw_ub!("invalid pointer for function lookup");
-        };
-        ret(self.prog.functions[func_name])
+    fn fn_from_ptr(&self, ptr: Value<M>) -> Result<Function> {
+        if let Value::Ptr(Pointer { thin_pointer: thin_ptr, .. }) = ptr {
```

Comment:
This should not ignore the matdata. It should require the metadata to be absent.